Add screenshot mocks

jkmassel opened the pull request:
Add support for screenshot mocks.

**To Test:**
- Ensure that all screenshots CI tasks pass

There's a _lot_ of code here – the majority of it is the mocks themselves, which aren't particularly interesting.

@shiki – this follows the pattern that you laid out previously, so I'm not sure if you want to take another look at it from an architectural standpoint? I think we'd talked about punting this all into a separate Swift package, which I'm ok to do if you'd like to, but I wasn't sure if it'd still apply given the move to Yosemite?

@rachelmcr and @JavonDavis  – you've previously worked on screenshots, so tagging you though you're both optional :) 

@woocommerce/release-managers – hoping at least one of you can take a detailed look through this (sorry!!)

jkmassel committed:
Add a basic initial object graph

jkmassel committed:
Add remotes that use an injected MockObjectGraph

The remote will return objects from the graph

jkmassel committed:
Add a mock authenticated state

jkmassel committed:
Allow mocking the network layer with a launch arg

jkmassel committed:
Move mocks into the Network layer

jkmassel committed:
Fix tests

shiki committed:
Prototype of overriding StorageManager

jkmassel committed:
Fix test launch

jkmassel committed:
Create mocks

jkmassel committed:
Fix broken single product screen

jkmassel committed:
Update screenshots to use mocks

jkmassel committed:
Add screenshots binary assets server

jkmassel committed:
Add product images support

jkmassel committed:
Remove Xcode time override

jkmassel committed:
Update snapshop

jkmassel committed:
Fix broken order merging

jkmassel committed:
Fix an issue with multiple mocks servers running

jkmassel committed:
Speed up tests slightly

jkmassel committed:
Update CocoaPods (and other dependencies)

jkmassel committed:
Use the 14.0 simulator

jkmassel committed:
Use Xcode 12 for screenshots

jkmassel committed:
Fix CircleCI with SwiftPM

jkmassel committed:
Fix up Github Action for screenshots

jkmassel committed:
Fix SPM support in CircleCI

jkmassel committed:
Fix GitHub actions issue

jkmassel committed:
Remove awscli command installation

jkmassel committed:
Add support for visitor order and revenue stats

jkmassel committed:
Add top products

jkmassel committed:
Make the status bar pretty

jkmassel committed:
Make promo screenshots in CI

jkmassel committed:
Fix broken tests

jkmassel committed:
Print the store path on launch

jkmassel committed:
Fix argument name

jkmassel committed:
Remove unneeded code

jkmassel committed:
Fix a bug causing visitor stats not to appear

jkmassel committed:
Remove unneeded abstraction

jkmassel committed:
Move ProcessInfo stuff into the BuildConfiguration

jkmassel committed:
Remove unneeded changes

jkmassel committed:
Add documentation for MockActionHandler helpers

jkmassel committed:
Spacing nitpick fix

jkmassel committed:
Re-sync version

AliSoftware commented on .github/workflows/screenshots.yml:
```diff
@@ -73,24 +63,13 @@ jobs:
+        bundler-cache: true
+
+    - name: Select Xcode Version
+      run: sudo xcode-select --switch /Applications/Xcode_12.app
```
I wonder if we shouldn't instead use [fastlane's `xcversion` action](https://docs.fastlane.tools/actions/xcversion/) – which automatically ensures the right version of Xcode is available on the machine and used by subsequent calls to `xcodebuild`?

That would make the Xcode version also enforced locally (not just on CI), and prevent the need of this (and of `sudo`) in CI.

jkmassel replied:
Haha I actually just learned that `xcversion` exists yesterday and didn't realize that it could be used here.

Changed in cfd0a24258937238af808732c8244279771aea31

AliSoftware commented on .github/workflows/screenshots.yml:
```diff
@@ -137,42 +108,65 @@ jobs:
+      - name: Install Native Dependencies
+        run: |
+          brew install imagemagick@6
+          brew link imagemagick@6 --force
```
nit: any reason why we need to use v6 here? (not that this PR should update to v7, as this can totally be done in a separate PR later if it's worth it rather than here; but was just wondering if you already looked at what currently prevents us from updating easily)

jkmassel replied:
Yeah, good question – I was in a rush and this worked fine. IIRC we've done some work on the release toolkit to support v7, but pinning the dependency in this way prevents it from breaking in a future v8 version some day too.

I agree this might be good to come back to, but yeah probably in a future PR!

AliSoftware replied:
Yeah I actually [updated release-toolkit to imagemagick 7 during my trial when I worked on Android Screenshots](https://github.com/wordpress-mobile/release-toolkit/pull/154) back in the day. Which means that the release toolkit itself does depend on imagemagick 7, so it might not be great to depend on 2 different versions for different projects sharing the toolkit…

I think we ultimately need to unify and migrate this to v7 to be aligned with release-toolkit, at least. But that could be done later if that's too much work to include in that PR and if we want to merge this sooner than later.

jkmassel replied:
Ah ok awesome – in that case I've bumped it to v7 in 3a58fd0161419487ae9f969fca94d7f0fb6bd11f – let's see how it works!

AliSoftware commented on .github/workflows/screenshots.yml:
```diff
@@ -137,42 +108,65 @@ jobs:
+      - name: Download Generated Screenshots
+        run: |
+          cd fastlane
+          aws s3 cp s3://$S3_BUCKET/2020-12-14-final/screenshots screenshots/  --recursive --exclude "*.html"
```
Will the run ID always be the same/unchanged? 🤔  (or is this a leftover of one of your debug session?)

jkmassel replied:
> is this a leftover of one of your debug session

Nice catch, thank you!!

Fixed in d6bbbefca7affdc570060b76c77cfcf4ab503420

AliSoftware commented on .github/workflows/screenshots.yml:
```diff
@@ -137,42 +108,65 @@ jobs:
+          name: raw-screenshots
+          path: fastlane/screenshots
+
+      - name: Install Promo Screenshot Fonts
```
YES! The risk with missing fonts for screenshots always bothered me in the past… Nice addition 👍

AliSoftware commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -10,6 +10,8 @@ import WordPressKit
 import WordPressAuthenticator
 import AutomatticTracks
 
+import class Yosemite.ScreenshotStoresManager
```
I like the targeted import there 👍

AliSoftware commented on WooCommerce/Classes/System/BuildConfiguration.swift:
```diff
@@ -32,4 +34,16 @@ enum BuildConfiguration: String {
     }
     #endif
+
+    static var shouldUseScreenshotsNetworkLayer: Bool {
+        return ProcessInfo.processInfo.arguments.contains("mocked-network-layer")
+    }
+
+    static var shouldLogoutAtLaunch: Bool {
+        ProcessInfo.processInfo.arguments.contains("logout-at-launch")
+    }
+
+    static var shouldDisableAnimations: Bool {
+        ProcessInfo.processInfo.arguments.contains("disable-animations")
+    }
```
Nit: At some point we should use Apple's recommended way to pass launch options instead – i.e. use `-mocked-network-layer` with a dash prefix then use `UserDefaults` to read them from the Argument Domain.
Not critical, but would be good practice.

jkmassel replied:
I would like this as well. One thing I wasn't clear on is whether that works inside of other packages (or bundles?), or just from the main one?

For now isolating this over here at least makes the call sites clean and we can improve it at some point

AliSoftware replied:
Yeah I like that you isolated them here 👌  will make it easier if we want to improve this later 👍 
Good question about cross-bundle 🤔  I don't have the answer but definitively something to double-check when we'll get to this!

AliSoftware commented on WooCommerce/WooCommerce.xcodeproj/xcshareddata/xcschemes/WooCommerceScreenshots.xcscheme:
```diff
@@ -1,7 +1,7 @@
 <Scheme
    LastUpgradeVersion = "1210"
-   version = "1.7">
+   version = "1.3">
```
Is this expected? 🤔

jkmassel replied:
Apparently - Xcode will change it back to `1.7` if you re-add a script phase, but only if it takes build arguments from a target.

Weird 🤷

AliSoftware commented on WooCommerce/WooCommerce.xcodeproj/xcshareddata/xcschemes/WooCommerceScreenshots.xcscheme:
```diff
@@ -11,24 +11,6 @@
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
       shouldUseLaunchSchemeArgsEnv = "YES">
-      <PreActions>
-         <ExecutionAction
-            ActionType = "Xcode.IDEStandardExecutionActionsCore.ExecutionActionType.ShellScriptAction">
-            <ActionContent
-               title = "Run Script"
-               scriptText = "IPHONE_TIME=$(ruby -e &quot;require &apos;time&apos;;puts Time.new(2007, 1, 9, 9, 41, 0).iso8601&quot;)&#10;&#10;xcrun simctl boot &quot;${TARGET_DEVICE_IDENTIFIER}&quot;&#10;&#10;xcrun simctl status_bar &quot;${TARGET_DEVICE_IDENTIFIER}&quot; override \&#10;    --time $IPHONE_TIME \&#10;    --dataNetwork wifi \&#10;    --wifiMode active \&#10;    --wifiBars 3 \&#10;  ...
-               <EnvironmentBuildable>
-                  <BuildableReference
-                     BuildableIdentifier = "primary"
-                     BlueprintIdentifier = "F997170123DBB97500592D8E"
-                     BuildableName = "WooCommerceScreenshots.xctest"
-                     BlueprintName = "WooCommerceScreenshots"
-                     ReferencedContainer = "container:WooCommerce.xcodeproj">
-                  </BuildableReference>
-               </EnvironmentBuildable>
-            </ActionContent>
-         </ExecutionAction>
-      </PreActions>
```
Doesn't this stop from using a mocked statusbar content (9:41, cellular icon, battery, etc) that we used to do for consistent screenshots?

jkmassel replied:
Fortunately no – this was moved into `fastlane`, so we just use `override_status_bar: true` in the `capture_ios_screenshots` and it does it all for us! 🎉

AliSoftware commented on WooCommerce/WooCommerceScreenshots/Utils/SnapshotHelper.swift:
```diff
@@ -38,22 +38,13 @@ func snapshot(_ name: String, timeWaitingForIdle timeout: TimeInterval = 20) {
-    case cannotFindHomeDirectory
     case cannotFindSimulatorHomeDirectory
-    case cannotAccessSimulatorHomeDirectory(String)
     case cannotRunOnPhysicalDevice
```
ℹ️ [Note for reviewers]: this whole file is not ours, it's provided by fastlane.

_(I lost some time reviewing it and trying to understand why some code unrelated to our use cases where changed in this PR, before realizing this)_

shiki replied:
Thanks!

AliSoftware commented on WooCommerce/WooCommerceScreenshots/WooCommerceScreenshots.swift:
```diff
@@ -54,6 +60,67 @@ class WooCommerceScreenshots: XCTestCase {
             .thenTakeScreenshot(named: "product-details")
     }
+
+    private let port = Int.random(in: 2048...65535)
```
I think we should put a mechanism in place to try the next port if the randomly-selected one fails.

For example, if it randomly selects port 12345 but that port is already used by some other app/process running on the machine, `server.start()` will fail with "port unavailable"; in those cases we should at least try the like the 5 next ports 12346...12350, until we find one working or bail.

jkmassel replied:
Yeah good call – went for a lazier solution in 2f483bc9cfb32938f8bfd2642539534193ef9f8f – passing `0` to the port number lets the kernel assign an open one 😜

AliSoftware replied:
TIL. 👍

AliSoftware commented on WooCommerce/WooCommerceScreenshots/WooCommerceScreenshots.swift:
```diff
@@ -54,6 +60,67 @@ class WooCommerceScreenshots: XCTestCase {
+
+    func handleWebRequest(env: [String: Any], startResponse: ((String, [(String, String)]) -> ()), sendBody: ((Data) -> ())) {
+        /// Extract the path, so http://localhost:9285/foo-bar would be `/foo-bar` here.
+        guard let path = env["PATH_INFO"] as? String else {
```
Always found it odd that Embassy used an untyped dictionary to access those values, as opposed to a type-checked structure, to expose those values; feels quite unnatural in Swift…

Maybe worth a comment to at least point to doc:
```suggestion
        // See https://github.com/envoy/Embassy#whats-swsgi-swift-web-server-gateway-interface for more about PATH_INFO
        guard let path = env["PATH_INFO"] as? String else {
```

jkmassel replied:
Yeah good call – addressed in 896a0404b8efdcf96458c9fb7bd7b8f42dee8f09 – I didn't use the suggestion so I could add a formatted comment which is the woo default

AliSoftware commented on WooCommerce/WooCommerceScreenshots/WooCommerceScreenshots.swift:
```diff
@@ -54,6 +60,67 @@ class WooCommerceScreenshots: XCTestCase {
+        }
+
+        /// Remove the leading `/`, so we're left with `foo-bar`
+        let assetName = path.replacingOccurrences(of: "/", with: "")
```
This feels a bit dangerous, especially if we start using namespaced assets at some point, e.g.
<img width="1140" alt="namespaced-xcasset" src="https://user-images.githubusercontent.com/216089/103890556-8f323880-50e8-11eb-9e08-b08e8fbc4afc.png">
(which then are referenced in the code by `UIImage(named: "icons/icon_add")` using their fully qualified name)

Maybe this could be safer instead:
```suggestion
        let assetName = path.replacingOccurrences(of: "^/", with: "", options: .regularExpression)
```

jkmassel replied:
Ah interesting ok!

Sounds good to me!

AliSoftware commented on WooCommerce/WooCommerceUITests/Screens/BaseScreen.swift:
```diff
@@ -20,6 +20,10 @@ class BaseScreen {
     @discardableResult
     func waitForPage() throws -> BaseScreen {
         XCTContext.runActivity(named: "Confirm page \(self) is loaded") { (activity) in
```
❤️  at the use of `XCTContext.runActivity(…)` 🙌

jkmassel replied:
That's all @rachelmcr! 😃 

I just copied it from WPiOS 😝

AliSoftware commented on Yosemite/Yosemite/Internal/String+Extensions.swift:
```diff
@@ -0,0 +1,22 @@
+    private static let slugSafeCharacters = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-")
+
+    public var slugified: String? {
+        guard let latin = self.applyingTransform(StringTransform("Any-Latin; Latin-ASCII; Lower;"), reverse: false) else {
```
❤️  ICU transforms FTW 👌

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockActionHandler.swift:
```diff
@@ -0,0 +1,56 @@
+        fatalError("Unable to handle action: \(action.identifier) \(String(describing: action))")
+    }
+
+    /// A helper for that immediately returns a success for `Result<Void>` closure patterns
```
nit:
```suggestion
    /// A helper that immediately returns a success for `Result<Void>` closure patterns
```

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockActionHandler.swift:
```diff
@@ -0,0 +1,56 @@
+        onCompletion(.success(()))
+    }
+
+    /// A helper for that immediately returns a success for `Error?` closure patterns
```
nit:
```suggestion
    /// A helper that immediately returns a success for `Error?` closure patterns
```

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockAppSettingsActionHandler.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    func setStatsVersionLastShown(siteId: Int64, statsVersion: StatsVersion) {
+        // TODO – we should persist this somewhere
```
nit: 💡 wdyt about using a warning to not forget about this later?
```suggestion
        #warning("TODO – we should persist this somewhere")
```

(And same for other `// TODO – xxx` comments I've seen in other mocks ofc)

jkmassel replied:
I've seen this pattern before, but we don't allow new warnings in PRs, so I can't use it 😅

If we were looking to fully build out the backend we'd persist it to a `plist` but it's not _required_. However, leaving the message and empty function body there hints that it's a future improvement we could make without requiring it or making a lot of noise 🤷‍♂️

AliSoftware replied:
Would then maybe `// TODO: xxx` (with the specific syntax that Xcode recognizes, i.e. exact `// TODO:` prefix) be better? Given that this:
1. is recognized by Xcode IDE
2. can be turned into a warning on demand [by a SwiftLint rule](https://realm.github.io/SwiftLint/todo.html) (there's even [another one that allows to put an expiration date on a TODO](https://realm.github.io/SwiftLint/expiring_todo.html) so the warning is only triggered after said date…)
 
That way we can keep this not being a warning by default… but be able to surface them as warning only on demand when handling maintenance 🤔

shiki replied:
I think we were also supposed to remove this `setStatsVersionLastShown` action. We used this to allow the user to change between v3 or v4 of the stats feature but we removed v3 a while ago. Only v4 is available. Maybe we just forgot to remove this. 🤷

shiki replied:
In any case, we only save these settings on files so if we need to save something, perhaps an `NSDictionary` is good enough since the screenshots flow don't really have a permanent state, right? It's just run and die?

jkmassel replied:
Addressed in 42f65e17c9df539736946c65e3c0a4f2155fe56f since yeah, there's no state. Just replaced with a no-op until it's removed!

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockOrderActionHandler.swift:
```diff
@@ -0,0 +1,48 @@
+            updater.upsert(orders)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: storage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
```
nit: felt off-balance that we needed to dispatch the completion to main here but not in the other cases. Wondering if it wouldn't be better to make `saveDerivedType` dispatch its own completion on main itself 🤔 

Probably not a big deal; but this surprised me when reading the code during review and I can see this dispatch easily be missed in other call sites when people don't realize this… 🤷

jkmassel replied:
Yeah, unfortunately that's code outside that scope of this PR. 

IMHO we should never emit changes on the main queue from a library function – the app layer should handle when it wants to bring the changes into the main queue to avoid random UI hitches. 

But it's [the pattern used elsewhere in this app](https://github.com/woocommerce/woocommerce-ios/blob/a1d8473bbaff62a6baa03e4f3a848dd0df4efb57/Yosemite/Yosemite/Stores/OrderStore.swift#L344-L346), which is why I copied it for this 🤷‍♂️

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/Helpers/ProductStockStatus+Mocks.swift:
```diff
@@ -0,0 +1,16 @@
+extension ProductStockStatus {
+    public static func from(quantity: Int64) -> ProductStockStatus {
+        switch quantity {
+            case Int64.min ..< 0:
+                return .onBackOrder
+            case 0:
+                return .outOfStock
+            case 1...Int64.max:
+                return .inStock
+            default:
+                preconditionFailure("Unable to parse quantity")
```
nit: you can leave out the extreme bounds. I find it clearer that way ;)

```suggestion
            case ..<0:
                return .onBackOrder
            case 0:
                return .outOfStock
            case 1...:
                return .inStock
            default:
                preconditionFailure("Unable to parse quantity")
```

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/MockObjectGraph.swift:
```diff
@@ -0,0 +1,481 @@
+// MARK: Product Accessors
+extension MockObjectGraph {
+
+    func product(forSiteId siteId: Int64, productId: Int64) -> Product {
+        return products(forSiteId: siteId).first { $0.productID == productId }!
+    }
```
Thinking out loud: seeing this force-unwrap I'm wondering: what's the real world app behavior when such productID doesn't exist, and wouldn't we want to be able to mock/test that "product not found in siteID" case in our tests? (iow return an optional here), like you do for `order(…)` and `review(…)` below? 🤔

jkmassel replied:
Yeah, good question – the product should be guaranteed to exist on the object graph here – we're just finding it in the list. Because `first` returns an optional, we force-unwrap it. If this ever fails there are bigger issues at play (ie – maybe we've loaded a production database in a test context?)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -537,6 +540,7 @@ ENV["APP_STORE_STRINGS_FILE_NAME"]="AppStoreStrings.pot"
       "app_store_screenshot-3"    => {desc: "app_store_screenshot_3.txt"},
       "app_store_screenshot-4"    => {desc: "app_store_screenshot_4.txt"},
       "app_store_screenshot-5"    => {desc: "app_store_screenshot_5.txt"},
+      "app_store_screenshot-6"    => {desc: "app_store_screenshot_6.txt"},
```
If we do have a new `app_store_screenshot_6.txt`, shouldn't `fastlane/screenshots.json` also have changed in that PR accordingly? I don't see the related change there in this branch though?

jkmassel replied:
Good call – this is no longer necessary :) 

Removed in 5c32829e2945b3654c49e0e5d23471270cf69698

AliSoftware reviewed (commented):
I only looked at the code itself in GitHub so far – didn't yet test the feature itself on my Mac – but figured I'd still post my initial review notes before going deeper and coming back to it to test it for real later.

Mostly nits, though a couple of questions too.

shiki commented:
I can take a look, possibly later today. 🙂

jkmassel committed:
Use xcversion instead of xcode-select

jkmassel committed:
Fix hardcoded screenshot set ID

shiki commented on .github/workflows/screenshots.yml:
```diff
@@ -137,42 +105,65 @@ jobs:
+
+      - name: Install Promo Screenshot Fonts
+        run: |
+          aws s3 cp s3://$S3_BUCKET/fonts.zip fonts.zip
```
Does this mean that we're downloading and installing the font every time the workflow is executed? Just curious, I really have no idea what's happening here. 🤪

jkmassel replied:
Yep that's exactly right, because we can't guarantee the Woo marketing fonts are on our CI machines, we need to grab them every time 🙃

jkmassel committed:
Remove app store screenshot 6

jkmassel committed:
Update Yosemite/Yosemite/Model/Mocks/Helpers/ProductStockStatus+Mocks.swift

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

shiki commented on .bundle/config:
```diff
@@ -0,0 +1,5 @@
+---
```
Somehow, I have this file in my local repo. Does this mean that we used to generate this?

AliSoftware replied:
It's not surprising: it's just because it used to be `.gitignore`'d (but now we want to commit it to have the same config for everyone).

I think the right move is to use the one from the PR replace the one you had locally. (This might make people have `git pull` error about the conflicting files once the PR is merged tho, might need to warn them in Slack)

shiki commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -318,11 +320,26 @@ private extension AppDelegate {
     func handleLaunchArguments() {
-        if ProcessInfo.processInfo.arguments.contains("logout-at-launch") {
+        if BuildConfiguration.shouldLogoutAtLaunch {
           ServiceLocator.stores.deauthenticate()
```
The indent is a bit off. 😉 Yay for nitpicking!

jkmassel replied:
I love nitpicks 😉

Addressed in f49e08d46f610cc4da670bd8a7ea6404194dd716

shiki commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -318,11 +320,26 @@ private extension AppDelegate {
+            /// Print the location of the core data DB for debugging
+            if let storeUrl = ServiceLocator
+                .storageManager
+                .persistentContainer
```
⚠️ I'd rather we don't use `persistentContainer` outside of `StorageManager`. I've been meaning to make this `private` because we couldn't mock `StorageManagerType` if this is still exposed. We made some progress towards that in https://github.com/woocommerce/woocommerce-ios/pull/2049 but we still have one more thing to do before we can make this private: https://github.com/woocommerce/woocommerce-ios/issues/1765.

Maybe we can expose the `storeURL` as a property of `StorageManagerType`?

jkmassel replied:
Works for me – definitely understand about isolating this!

I personally like printing the path the app's documents directory or something on startup, because when debugging I like to open up the core data DB a bunch to see what it's doing. 

But for now, I'm fine to just drop this code to avoid further changes – I'll replace it some other future tooling thing probably 😜

Removed in 046c07d7a91a0dcb5acdd438a8f2909a31421c31

shiki replied:
Removal was a better decision than anything I could have imagined. 🤣

shiki commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -318,11 +320,26 @@ private extension AppDelegate {
+                debugPrint("💿 \(storeUrl.path)")
+            }
+
+            ServiceLocator.setStores(ScreenshotStoresManager(storageManager: ServiceLocator.storageManager))
```
🎉 🎉 🎉

jkmassel committed:
Use safer URL replacement

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

shiki commented on WooCommerce/WooCommerceScreenshots/WooCommerceScreenshots.swift:
```diff
@@ -54,6 +60,67 @@ class WooCommerceScreenshots: XCTestCase {
+    }
+
+    func handleWebRequest(env: [String: Any], startResponse: ((String, [(String, String)]) -> ()), sendBody: ((Data) -> ())) {
+        /// Extract the path, so http://localhost:9285/foo-bar would be `/foo-bar` here.
```
It'd be great if there's a summary of what we're doing here and why we need to do it. I admittedly don't understand what's going on. 😅

JavonDavis replied:
+1

jkmassel replied:
Added a bit more info in 896a0404b8efdcf96458c9fb7bd7b8f42dee8f09 – hope it's helpful – happy to add more if needed!!!

shiki commented on Yosemite/Yosemite/Internal/String+Extensions.swift:
```diff
@@ -0,0 +1,22 @@
+
+    private static let slugSafeCharacters = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-")
+
+    public var slugified: String? {
```
I believe this can be module-level only. 

```suggestion
    var slugified: String? {
```

shiki commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockActionHandler.swift:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /// A helper for saving mock objects into Core Data
+    func save<T, U>(mocks: [T], as dataType: U.Type, onCompletion: @escaping (Error?) -> ()) where U: ReadOnlyConvertible & NSManagedObject {
```
Noice!

jkmassel committed:
Use imagemagick 7

shiki commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockNotificationCountActionHandler.swift:
```diff
@@ -0,0 +1,28 @@
+        onCompletion(objectGraph.currentNotificationCount)
+    }
+
+    func reset(siteId: Int64, type: Note.Kind, onCompletion: () -> Void) {
+        // Don't do anything when resetting
+    }
```
I guess this is a disadvantage of `unimplementedAction(action: action)`. We're forced to add `case` for everything instead of just excluding them and letting `default` do nothing. 🙂

jkmassel replied:
Ya there's definitely a trade-off there – I'd prefer the screenshots code crash rather than silently ignore signals that the app emits and create screenshots with missing data 😅

shiki commented on Yosemite/Yosemite/Model/Mocks/Graphs/ScreenshotsObjectGraph.swift:
```diff
@@ -0,0 +1,366 @@
+    static var shippingAddress: Address {
+        return billingAddress
+    }
+}
```
ℹ️ This does not seem to be used. It's either that or Xcode is lying to me. 😄 

```suggestion

```

jkmassel replied:
Ah yep – I ended up just using a `struct` for this but forgot to remove it.

Thanks! Addressed in 23a0b02ebc2f8d5b7b017f8960db6f3681f9d35a

(I'm too spoiled by Rust, which complains about this kind of thing!!)

shiki replied:
Xcode should complain about these things too. Just saying. 🤷

shiki commented on Yosemite/Yosemite/Model/Mocks/Graphs/ScreenshotsObjectGraph.swift:
```diff
@@ -0,0 +1,366 @@
+
+struct ScreenshotObjectGraph: MockObjectGraph {
+
+     public let userCredentials = Credentials(
```
ℹ️ This and all the other `public` properties and methods below do not need to be `public`. 🙂 

```suggestion
    let userCredentials = Credentials(
```

jkmassel replied:
Yeah – I had these in place in case we were moving to to a package, but I'll go back and audit visibility when we do that in a future PR :) 

Addressed in c99bc531f92a9b1b4fdbb3ba6e5b9eb5efc108b0

shiki commented on Yosemite/Yosemite/Model/Mocks/Graphs/ScreenshotsObjectGraph.swift:
```diff
@@ -0,0 +1,366 @@
+
+    /// The possible value of an order when generating random stats
+    ///
+    let orderValueRange = 100 ..< 500
```
ℹ️  These can be `private`. 🙂 

```suggestion
    /// The probability of a sale for each visit when generating random stats
    ///
    private let orderProbabilityRange = 0.1...0.5

    /// The possible value of an order when generating random stats
    ///
    private let orderValueRange = 100 ..< 500
```

jkmassel replied:
Sure can – also addressed in c99bc531f92a9b1b4fdbb3ba6e5b9eb5efc108b0

shiki commented on Yosemite/Yosemite/Model/Mocks/MockObjectGraph.swift:
```diff
@@ -0,0 +1,481 @@
+            end: Date.from(dateString: end)
+        )
+    }
+}
```
ℹ️ This seems unused. 🙂 

```suggestion

```

jkmassel replied:
Removed in 2a5fa077afbed46e6d31b1ec318fb523d1b27c03

shiki commented on Yosemite/Yosemite/Model/Mocks/MockObjectGraph.swift:
```diff
@@ -0,0 +1,481 @@
+        formatter.dateFormat = "yyyy-MM-dd"
+        return formatter.date(from: dateString)!
+    }
+}
```
Do you think you'll be able to use some of the formatters in [`DateFormatter+Woo`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Extensions/DateFormatter+Woo.swift)? I'm not sure. 🤔 I'm just sharing in case you didn't know about that.

jkmassel replied:
Nice – I wasn't aware those existed. For creating these mock objects I still prefer the fluent approach, but calling out to those formatters avoids reinventing this particular wheel and will also beneficially cause the code to break if we move to a different versioning scheme (say in a future statsv5?)

Addressed in 2a5fa077afbed46e6d31b1ec318fb523d1b27c03

shiki reviewed (changes requested):
Hi, @jkmassel! I like how this turned out! 🎉 

I'm requesting changes because of one little tiny thing: the usage of `persistentContainer`. Please see the inline comment marked with ⚠️ . The others are just suggestions or nitpicks. 😄 

> I think we'd talked about punting this all into a separate Swift package, which I'm ok to do if you'd like to, but I wasn't sure if it'd still apply given the move to Yosemite?

Yeah, I'm not sure either. The only concern here is that there are lots of classes/structs in Yosemite now so maybe it makes sense to have a separate Swift package. That should make it easier for product developers to grok Yosemite. So maybe it's worth a try? 

I'm not sure how much effort is involved in that. We might have some surprising things that need to be done. 😬

AliSoftware commented on .bundle/config:
```diff
@@ -0,0 +1,5 @@
+BUNDLE_RETRY: "3"
+BUNDLE_PATH: "vendor/bundle"
+BUNDLE_JOBS: "3"
+BUNDLE_WITHOUT: "screenshots"
```
This made me realize a side effect of this: when we want to generate screenshots, we will run `bundle install --with screenshots` – or, as it's nowadays suggested by bundler instead, `bundle config --local --delete without` (see [this in WPAndroid for example](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/.circleci/config.yml#L460-L462)).

In both cases this will **have the effect of modifying this `bundle/config` file** to remove this `BUNDLE_WITHOUT`, **a change which should NOT be commited** (as we don't want everyone to install gems that are useless for most people if they never have to create screenshots in most cases).

This is probably unlikely to happen as most devs will never have to do that if screenshots are generated by an MIE or on CI, but something to keep in mind still…

jkmassel replied:
Yeah, this is true. It's kinda suboptimal no matter which way you go – we don't want people having to remember to run `bundle install --without screenshots` (or whatever the new syntax is 😅) 

But we also don't want to dirty the file every time. 

IMHO this is the _best_ option because as you note it'll only affect MIEs and not everybody else 😄

Once we drop ruby for all the things and only MIEs are using it, IMHO we could just remove the `without` part and make it a requirement if you want to use Ruby?

JavonDavis commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockProductActionHandler.swift:
```diff
@@ -0,0 +1,72 @@
+            productStore.upsertStoredProducts(readOnlyProducts: products, in: storage)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: storage) {
```
@jkmassel Just a quick question, it's probably not super important here either but could we end up a race condition with the perform block here? I'm wondering if _technically_ we could end up saving before the perform block completes

jkmassel replied:
Ya this is an excellent point – I can't believe I missed this 🙈 

It's the same code as is used here: 

https://github.com/woocommerce/woocommerce-ios/blob/bdbabd20ca3ee0a581c7c1b1e5c8bafd7a319eeb/Yosemite/Yosemite/Stores/ProductStore.swift#L345-L357


@shiki – WDYT? IMHO this makes sense to fix in both places right?

shiki replied:
I have been confused by this in the past. This is fine as far as I can tell. The `saveDerivedType` has a `perform` call too. 🙂 So they should happen in sequence. 

https://github.com/woocommerce/woocommerce-ios/blob/bdbabd20ca3ee0a581c7c1b1e5c8bafd7a319eeb/Storage/Storage/CoreData/CoreDataManager.swift#L143-L152

<img src="https://media.giphy.com/media/QMHoU66sBXqqLqYvGO/giphy.gif" width="300">

jkmassel committed:
Remove unused code

jkmassel committed:
Address method and property visiblity feedback

jkmassel committed:
Use pre-existing date formatters

jkmassel committed:
Add xcode-install gem

jkmassel committed:
Remove duplicated build step

jkmassel committed:
Use xcversion when taking screenshots too

jkmassel committed:
Bump simulator version to 14.3

jkmassel committed:
Remove TODO comments

AliSoftware commented:
Moving this to milestone 5.9 as I'm code-freezing 5.8 today

shiki commented:
@jkmassel Let me know when this is ready for another round!

jkmassel committed:
Let the kernel pick the server port

jkmassel committed:
Fix indentation

jkmassel committed:
Remove storeUrl printing

Per https://github.com/woocommerce/woocommerce-ios/pull/3337#discussion_r553483361

jkmassel committed:
Fix slugified visibility

Co-authored-by: Jayson Basañes <jayson@basanes.net>

jkmassel committed:
Add additional comments for built-in web server

jkmassel committed:
Make hound happy

jkmassel commented:
@shiki, @AliSoftware, and @JavonDavis – thanks for the excellent feedback. I'm _pretty_ sure I've addressed all of it in this PR (apologies if I missed one, there was a lot!! 😁)

Ready for another look at your convenience! 🙇

shiki reviewed (approved):
Thanks, @jkmassel! 

The tests didn't run for me on iPhone 11. It just stayed at “Running tests...”. It works on iPhone 12 though. Not sure if this is the expected behavior so I thought I'd share.

AliSoftware commented on WooCommerce/WooCommerceScreenshots/ScreenshotImages.xcassets/akoya-pearl-shades.imageset/Contents.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "images" : [
+    {
+      "filename" : "Mask-Group-2.png",
```
[super-nitpick] The OCD in me is kicking in when I see the name of the PNG file that is completely different / not based on the basename of the `.imageset` 😅  (ditto for other imagesets below)

jkmassel replied:
Hah nice! Fixed in c52879a4b911a43c2b645309b9f5f893c38ad6d2 😂

AliSoftware commented on Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockStatsActionV4Handler.swift:
```diff
@@ -0,0 +1,61 @@
+            case .thisMonth: success(onCompletion)
+            case .thisYear:
+                store.upsertStoredOrderStats(readOnlyStats: objectGraph.thisYearOrderStats, timeRange: timeRange)
+                onCompletion(nil)
```
My OCD speaking again: feels a bit unbalanced and harder to read to have some `case` having their action on the same line but others (just the last one here) being on separate lines (because there's more than one expression).

At first skim my mind read it as if that was a C-style switch (i.e. with implicit `fallthrough`) and thought that all 4 cases were executing the last 2 lines (`store.upsert…` + `onCompletion`), before I clicked back into realizing there were some `success(onCompleteion)` calls hiding there.

It might only be a nit but even if it would make the code spread a bit more vertically, I think it would be worth always putting the instructions after each case on their own lines. In this specific example, given the instructions are the same for the 3 first cases, we could DRY the 3 cases on one line though:

```suggestion
            case .today, .thisWeek, .thisMonth:
                success(onCompletion)
            case .thisYear:
                store.upsertStoredOrderStats(readOnlyStats: objectGraph.thisYearOrderStats, timeRange: timeRange)
                onCompletion(nil)
```

Ditto for all other similar `switch` in the same file below ofc.

jkmassel replied:
Yeah fair enough – fixed all of the cases in this file in 11e4afa0a6e8ae757a51e5a501f8f4b2b50a8886 and verified that there aren't others in the other files.

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -699,7 +708,7 @@ ENV["APP_STORE_STRINGS_FILE_NAME"]="AppStoreStrings.pot"
 
   def ios_locales
-    "ar de-DE es-ES fr-FR he id it ja ko nl-NL pt-BR ru sv tr zh-Hans zh-Hant en-US".split()
+    "ar de-DE en-US es-ES fr-FR he id it ja ko nl-NL pt-BR ru sv tr zh-Hans zh-Hant".split()
```
Ah, one thing my OCD likes very much! 😄

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -471,7 +477,10 @@ ENV["APP_STORE_STRINGS_FILE_NAME"]="AppStoreStrings.pot"
-      dark_mode: options[:mode].to_s.downcase == "dark"
+      dark_mode: options[:mode].to_s.downcase == "dark",
+
+      # Make the status bar pretty
```
💄 😊 ✨

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -415,6 +418,9 @@ ENV["APP_STORE_STRINGS_FILE_NAME"]="AppStoreStrings.pot"
   lane :take_screenshots do | options |
 
+    # Ensure we're using the right version of Xcode
+    xcversion(version: "~> 12.0")
```
❤️

AliSoftware reviewed (approved):
LGTM!

Only left a couple of nits that could be nice to address before merge, but feel free to ignore them.

jkmassel commented:
> The tests didn't run for me on iPhone 11

@shiki – thanks for pointing this out. Did you happen to run it again after shutting off the simulator? I've tested on iPhone 11 across a few iOS versions and they all seem to work. My guess would be that some kind of Xcode => Simulator XPC thing was having a bad day and that's why it didn't work? We don't use iPhone 11 for testing, so if we run into this in the future, I think it'd be ok to address it then?

jkmassel committed:
Make switch statements more readable

jkmassel committed:
Fix image names in asset catalogue

jkmassel committed:
Merge remote-tracking branch 'origin/develop' into add/screenshot-mocks

# Conflicts:
#	Yosemite/Yosemite.xcodeproj/project.pbxproj

jkmassel commented:
Force-pushed to fix a failed merge with `develop` – merging once it turns 🟢 

Thanks for the reviews y'all!! 🙌

shiki commented:
> Did you happen to run it again after shutting off the simulator?

Yeah, I tried everything that I know including clearing derived data and stuff. I think it just hates me. 

But yeah, I wouldn't worry about it.

jkmassel committed:
Fix missing case post-merge
